beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -30,24 +32,38 @@ enum RangeBlockDataRequest<E: EthSpec> {
             DataColumnsByRangeRequestId,
             ByRangeRequest<DataColumnsByRangeRequestId, DataColumnSidecarList<E>>,
         >,
+        /// The column indices corresponding to the request
+        column_peers: HashMap<DataColumnsByRangeRequestId, Vec<u64>>,
         expected_custody_columns: Vec<ColumnIndex>,
     },
 }
 
+#[derive(Debug)]
+pub struct CouplingError {
+    pub(crate) msg: String,
+    pub(crate) column_and_peer: Option<(Vec<(u64, PeerId)>, PeerAction)>,
```

Comment:
Just a note here that we don't track the block root that had missing columns, so we just request the column for the entire range - might be something to optimise for when we get to higher blob counts.

for 48 blobs, each column is 96kb, so in the worst case if we're missing a single column for a block, we'd end up downloading the entire batch (96kb * 32 = 3072kb).

However, this is also an edge case - in normal cases, peer usually not return the columns or returns everything. The current logic handles this case fine, just to note a possible future optimisation. I think some clients are already using byRoots to retrieve missing columns, but in the current structure this feels pretty complicated to do.

Reply:
Agreed